crates/common/tedge_config_macros/examples/macro.rs

```diff
@@ -0,0 +1,178 @@
+use camino::Utf8PathBuf;
+use std::net::IpAddr;
+use std::net::Ipv4Addr;
+use std::num::NonZeroU16;
+use std::path::PathBuf;
+use tedge_config_macros::*;
+
+static DEFAULT_ROOT_CERT_PATH: &str = "/etc/ssl/certs";
+
+define_tedge_config! {
+    device: {
+        #[tedge_config(readonly(
+            write_error = "Device id is inferred from the certificate",
+            function = "device_id",
+        ))]
+        #[doku(as = "String")]
+        id: Result<String, String>,
+
+        /// Path where the device's private key is stored
+        #[tedge_config(example = "/etc/tedge/device-certs/tedge-private-key.pem", default(function = "default_device_key"))]
+        #[doku(as = "PathBuf")]
+        key_path: Utf8PathBuf,
+
+        /// Path where the device's certificate is stored
+        #[tedge_config(example = "/etc/tedge/device-certs/tedge-certificate.pem", default(function = "default_device_cert"))]
+        #[doku(as = "PathBuf")]
+        cert_path: Utf8PathBuf,
+
+        /// The default device type
+        #[tedge_config(example = "thin-edge.io")]
+        #[tedge_config(rename = "type")]
```

Comment:
@jarhodes314 The need for a specific `#[tedge_config(rename)]`  vs `#[serde(rename)]` makes sense.

About `alias`, I think the confusion is beyond just a lack of documentation. It's not obvious that `#[serde(alias = ...)]` has a specific meaning for the `tedge_config`  and that this meaning is added thanks to `doku`.

So if things were free I would vote for `#[tedge_config(deprecated_name = "cafile")]`.

Reply:
I've now modified this to match your suggestion. I've also added `#[tedge_config(deprecated_key = "mqtt.port")]`. The difference between `deprecated_name` and `deprecated_key` is that `deprecated_name` is for renaming a field/group, but `deprecated_key` is for mapping one key to a different structure (e.g. `mqtt.port` -> `mqtt.bind.port`), which cannot just translate to `#[serde(alias)]`. The macro will emit compiler errors if the user uses the wrong one (by checking if the input contains `.`, if it does, it can only be a key, and if it doesn't, it can only be a (field/group) name as all fields are nested in at least one group).